crates/bevy_ecs/src/system/query.rs

```diff
@@ -2097,7 +2097,21 @@ pub struct QueryLens<'w, Q: QueryData, F: QueryFilter = ()> {
 
 impl<'w, Q: QueryData, F: QueryFilter> QueryLens<'w, Q, F> {
     /// Create a [`Query`] from the underlying [`QueryState`].
-    pub fn query(&mut self) -> Query<'w, '_, Q, F> {
+    pub fn query(&mut self) -> Query<'_, '_, Q, F> {
```

Comment:
So the new flow is
1) `query` -> `Query<'w, s'>`
2) `lens = &'a query.as_query_lens()` -> `QueryLens<'a>`
3) `lens.query()` -> `Query<'a, 'a>`
4) `get_inner()` -> `QueryItem<'a>`

(and the old one is
3) `lens.query()` -> `Query<'w, 'a>`
4) `get_inner()` -> QueryItem<'w>`
)

So in the old case, the item we get only depends on the 'w lifetime, so the compiler doesn't stop us from creating a second object with the 'a lifetime via `lens.query()`.

I think the change makes sense; out of curoisity, would this PR https://github.com/bevyengine/bevy/pull/15396 also work to fix your compile-fail test?
Since the QueryItem we get now also depends on the 'a lifetime

Reply:
I did't completely follow your notation, but your conclusion sounds right!  We had an `&'a mut QueryLens<'w>`, and currently that gives a `Query<'w, 'a>` which we can use with `get_inner()` to get a `QueryItem<'w>`, but with this change it gives a `Query<'a, 'a>` that we can only use to get a `QueryItem<'a>`.  

I believe the compile-fail test would still compile with #15396, because the compiler would see that the concrete type of `QueryItem<'w, 'a>` is `Mut<'w, Foo>`, which doesn't involve the `'a` lifetime.